### ISSUE TITLE
Problem performing a readLocation call using the M2XStreamClient library

### ISSUE THREAD

**Issue by kristinpeterson** [attm2x/m2x-arm-mbed#5]
Reported by @jflynn129

Using client code hosted here: https://developer.mbed.org/teams/ATT-M2X-team/code/M2XStreamClient/

Problem: When performing a readLocation call using the M2XStreamClient
library, the call hangs and never returns.

Investigation:
I have a large amount of data in my Location store on M2X
(https://m2x.att.com/devices/e83cdd8645ab1a7c0c480156efbf78f6).  When I
call the readLocation method, and it never returns.
Problem:
The problem is when readLocation is called, it expects a
“Content-Length:” header in the M2X server response.  But the M2X
response does not include this header, so as a result, the program just
hangs.  A copy of the M2X response is below.
Possible Solutions:
1. Insert a read-timeout in the ‘readContentLength’ function so it
doesn’t hang forever.
2. Check with the M2X server guys and confirm that this response is
correct (no Content-Length header).  It may be a problem in their code.
## Response received that is missing “Content-Length”

Content-Type: application/json; charset=UTF-8
Date: Sat, 10 Sep 2016 21:53:34 GMT
Server: nginx
Vary: Accept-Encoding
Vary: Accept
X-M2x-Version: v2.99.2

{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T21:53:33.817Z","waypoints":[{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T21:53:33.817Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T21:50:01.028Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T21:40:38.061Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T21:31:31.009Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T21:27:54.698Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T21:18:11.895Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T21:13:54.558Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T21:11:02.413Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T21:06:45.441Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T20:51:06.037Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T20:46:14.902Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T20:41:14.399Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T20:39:18.270Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T20:33:27.345Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T20:25:15.459Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T20:17:32.574Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T20:12:25.781Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T20:03:58.228Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T20:01:00.651Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T19:59:08.666Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T19:56:11.988Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T19:47:36.097Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T19:02:38.699Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T19:00:39.447Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T18:58:31.856Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T18:54:17.569Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T18:50:55.032Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T18:44:47.682Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T18:43:20.067Z"},{"name":"Wake
Forest","latitude":33.0079,"longitude":-96.7516,"elevation":"697","timestamp":"2016-09-10T18:40:36.892Z"}]}

**Comment by kristinpeterson** [comment]
@xxuejie one more mbed issue to investigate, when you have a moment.

**Comment by xxuejie** [comment]
@kristinpeterson Thanks for creating the issue! I need to think about *if* we can work without a `Content-Length`, however in the meantime, do you think we can relay the information to the M2X server team to see how they thinks here?

**Comment by jemc** [comment]
For whatever it's worth, when I call `GET /v2/devices/:id/location` or `GET /v2/devices/:id/location/waypoints` on one of my own devices, I get a response with `Content-Length` included each time.  Has anyone been able to reproduce the missing `Content-Length` header?

**Comment by jflynn129** [comment]
Joe, try with the device in the report above and see if it operates as described.  Maybe the device isn't created correctly (?) but I don't know why it would be.  I had posted a lot (several hours worth) of location updates before this started happening.

**Comment by jemc** [comment]
@jflynn129 just tried both API calls with the device id specified in the ticket description.  I still see a `Content-Length` header.  Are you able to reliably reproduce, or does it happen only sometimes?

Note that I am simply calling the M2X API from my terminal, not using the arm-mbed library, in case that is a factor in reproducing the problem.

**Comment by jflynn129** [comment]
Interesting, I am only using the arm-mbed library, haven't tried it from
a terminal (honestly, I don't know how to do it from a terminal).  When
I call it from the arm-mbed library, it fails every time, I get stuck in
readContentLength (watched it from ULINK).  Just curious, can you send
me a copy of the output you got so I can compare it to mine.

**Comment by jemc** [comment]
I found that I was seeing `Content-Length` because my HTTP client ([HTTPie](https://httpie.org/)) was using `Accept-Encoding: gzip, deflate`, so the server was sending `Content-Encoding: gzip`, which is always length-specified.

If I used `curl` instead, `Content-Encoding: gzip` wasn't used and I got `Transfer-Encoding: chunked` instead (which uses no `Content-Length`).

However, if I forced HTTP 1.0 (with the `--http1.0` flag) to mimic the arm-mbed library, I got neither `Content-Length` nor `Transfer-Encoding: chunked`.  I should have received `Content-Length`, since `Transfer-Encoding: chunked` is not allowed in HTTP 1.0.

We'll be looking into this further.

**Comment by jflynn129** [comment]
I ran into this issue again; getting a HTTP 1.1 with Transfer-Encoding:
chunked

Were you ever able to resolve this issue?  This is being returned from the
AT&T M2X site...

On Tue, Sep 13, 2016 at 11:39 AM, Joe Eli McIlvain <notifications@github.com

**Comment by kristinpeterson** [comment]
@jemc could you please provide an update on this when you have a moment?

**Comment by jflynn129** [comment]
The AT&T M2X site is sending a message with an “HTTP 1.1” and “Transfer-Encoding: chunked”  header.  Previously, it was determined that there was a problem with the handling of Content-Length.  I was previously able to avoid this by reverting the header to HTTP 1.0 so the problem went away.  Now that the M2X site is returning this header, it is again a problem.  

It was mentioned (Joe Mcllvain) that this is an HTTP 1.0 bug limitation resulting from the underlying technology used by M2X.

Specifically, a web server called Unicorn.  Where HTTP 1.0 requests return as (invalid) HTTP 1.1 responses when the response body is longer than a given threshold. 

Possible options forward for resolving would be:
- Migrate to use a different web server other than Unicorn - one with proper support for HTTP 1.0.
- Coordinate with the upstream maintainers of Unicorn to patch the bug and fix support for HTTP 1.0.
- Discontinue M2X support for HTTP 1.0 and update the mbed library (and other client libraries) to use HTTP 1.1.

I have no opinion as to the best approach, I am simply using the M2X libraries as they are implemented and distributed.  So long as the libraries that are distributed work with the M2X service, I will be happy :-)

**Comment by kristinpeterson** [comment]
@jflynn129 thank you for reporting this issue! We have researched this internally and have a plan in place to remedy this for all clients/devices that require HTTP 1.0. 

The fix will be implemented within the M2X core architecture - stay tuned to the [M2X Changelog](http://changelog-m2x.att.com/) or [subscribe](http://eepurl.com/b2SNr5) to receive email updates to discover when the change goes live!

Closing this ticket as no changes are required on the `m2x-arm-mbed` client to resolve this issue, a ticket has been filed in M2X.
